tests/integration/framework/process/exec/exec.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2023 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package exec
+
+import (
+	"context"
+	"io"
+	oexec "os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/dapr/dapr/tests/integration/framework/process/exec/iowriter"
+	"github.com/dapr/dapr/tests/integration/framework/process/exec/kill"
+)
+
+type options struct {
+	stdout io.WriteCloser
+	stderr io.WriteCloser
+
+	runErrorFn func(error)
+	exitCode   int
+}
+
+type Option func(*options)
+
+type exec struct {
+	lock sync.Mutex
+	cmd  *oexec.Cmd
+
+	args       []string
+	binPath    string
+	runErrorFn func(error)
+	exitCode   int
+	stdoutpipe io.WriteCloser
+	stderrpipe io.WriteCloser
+}
+
+func New(t *testing.T, binPath string, args []string, fopts ...Option) *exec {
+	t.Helper()
+
+	defaultExitCode := 0
+	if runtime.GOOS == "windows" {
+		// Windows returns 1 when we kill the process.
+		defaultExitCode = 1
+	}
+
+	opts := options{
+		stdout: iowriter.New(t, filepath.Base(binPath)),
+		stderr: iowriter.New(t, filepath.Base(binPath)),
+		runErrorFn: func(err error) {
+			t.Helper()
+			if runtime.GOOS == "windows" {
+				// Windows returns 1 when we kill the process.
+				assert.ErrorContains(t, err, "exit status 1")
+			} else {
+				assert.NoError(t, err, "expected %q to run without error", binPath)
+			}
+		},
+		exitCode: defaultExitCode,
+	}
+
+	for _, fopt := range fopts {
+		fopt(&opts)
+	}
+
+	return &exec{
+		binPath:    binPath,
+		args:       args,
+		stdoutpipe: opts.stdout,
+		stderrpipe: opts.stderr,
+		runErrorFn: opts.runErrorFn,
+		exitCode:   opts.exitCode,
+	}
+}
+
+func (e *exec) Run(t *testing.T, ctx context.Context) {
+	t.Helper()
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	t.Logf("Running %q with args: %s %s", filepath.Base(e.binPath), e.binPath, strings.Join(e.args, " "))
+
+	//nolint:gosec
+	e.cmd = oexec.CommandContext(ctx, e.binPath, e.args...)
+
+	e.cmd.Stdout = e.stdoutpipe
+	e.cmd.Stderr = e.stderrpipe
+
+	require.NoError(t, e.cmd.Start())
+}
+
+func (e *exec) Cleanup(t *testing.T) {
+	t.Helper()
+	e.lock.Lock()
```

Comment:
can this be executed more than once? just thinking that if you kill the command on the first iteration, all the next ones will fail checkExit failing the test. or maybe it should be protected to run once only?

Reply:
`exec` is a process for executing a program file, and is more of a helper for the other processes which want to run binaries, as of this PR, daprd, placement, and sentry. For each test case, the framework will  `Run` and `Cleanup` each process exactly once, and the binary processes are going to do the same with their exec. Processes are not intended to be run more than once.. I can add a check for this if you think this will be beneficial?

Reply:
I meant the Cleanup method, just wondering if there was a path where this would be called more than once and maybe failing the test at the end even if it was fine.  If not then don't worry

Reply:
Sure, I can refactor processes a bit to ensure these functions only get called once. This may be beneficial for future developers.